astropy/io/ascii/ui.py

```diff
@@ -627,6 +627,37 @@ def _guess(table, read_kwargs, format, fast_reader):
         cparser.CParserError,
     )
 
+    # Determine whether we should limit the number of lines used in the guessing.
+    # Note that this does not currently work for file objects, so we set this to
+    # False if a file object was passed in.
+    from astropy.io.ascii import conf  # avoid circular imports
+
+    limit_lines = conf.guess_limit_lines and not hasattr(table, 'read')
+
+    # Don't limit the number of lines if there are fewer than this number of
+    # lines in the table. In fact, we also don't limit the number of lines if
+    # there are just above the number of lines compared to the limit, up to a
+    # factor of 2, since it is fast to just go straight to the full table read.
+    table_guess_subset = None
+
+    if limit_lines:
+
+        if isinstance(table, list):
+            if len(table) > 2 * limit_lines:
+                table_guess_subset = table[:limit_lines]
+        else:
```

Comment:
```suggestion
        elif isinstance(table, io.BufferedReader):
            line_len = len(table.readline())
            table.seek(0)
            table_guess_subset = table.readlines(line_len * limit_lines)
            table.seek(0)
        else:
```
Kind of a crude guess; not sure if this would have to be wrapped in `get_readable_fileobj` on top.

Reply:
I'd rather leave supporting file objects to another PR - in particular I need to understand if some file objects are streaming and don't support ``seek(0)`` for instance.

Reply:
Makes sense. `get_readable_fileobj` does check this and I believe, is making sure the returned file object provides `seek(0)`, but at that point may already have created more overhead than directly reading the file in.
https://github.com/astropy/astropy/blob/3099d86fbf0f7f88ed079fd50a563ad661fa2ca1/astropy/utils/data.py#L376-L386